neat_cache/lib/src/providers/redis.dart

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:io' show SocketException;
+import 'package:dartis/dartis.dart';
+import '../../cache_provider.dart';
+
+class _RedisContext {
+  final Connection connection;
+  final Client client;
+  final Commands<String, List<int>> commands;
+  _RedisContext({this.connection, this.client, this.commands});
+}
+
+typedef _Op<T> = Future<T> Function(Commands<String, List<int>>);
+
+class RedisCacheProvider extends CacheProvider<List<int>> {
+  final String _connectionString;
+  final Duration _connectTimeLimit;
+  final Duration _commandTimeLimit;
+  final Duration _reconnectDelay;
+
+  Future<_RedisContext> _context;
+  bool _isClosed = false;
+
+  RedisCacheProvider(
+    this._connectionString, {
+    Duration connectTimeLimit = const Duration(seconds: 30),
+    Duration commandTimeLimit = const Duration(milliseconds: 200),
+    Duration reconnectDelay = const Duration(seconds: 30),
+  })  : _connectTimeLimit = connectTimeLimit,
+        _commandTimeLimit = commandTimeLimit,
+        _reconnectDelay = reconnectDelay {
+    assert(_connectionString.isNotEmpty, 'connectionString must be given');
+    assert(!_connectTimeLimit.isNegative, 'connectTimeLimit is negative');
+    assert(!_commandTimeLimit.isNegative, 'commandTimeLimit is negative');
+    assert(!_reconnectDelay.isNegative, 'reconnectDelay is negative');
+  }
+
+  Future<_RedisContext> _createContext() async {
+    try {
+      print('### CONNECTING!!!');
+      final connection = await Connection.connect(_connectionString)
+          .timeout(_connectTimeLimit);
+      // Create context
+      final client = Client(connection);
+      return _RedisContext(
+        connection: connection,
+        client: client,
+        commands: client.asCommands(),
+      );
+    } on TimeoutException {
+      throw IntermittentCacheException('connect failed with timeout');
+    } on SocketException {
+      throw IntermittentCacheException('connect failed with socket exception');
+    } on Exception {
+      throw IntermittentCacheException('connect failed with exception');
+    }
+  }
+
+  Future<_RedisContext> _getContext() {
+    if (_context != null) {
+      return _context;
+    }
+    _context = _createContext();
+    scheduleMicrotask(() async {
+      _RedisContext ctx;
+      try {
+        ctx = await _context;
+      } on IntermittentCacheException {
+        // If connecting fails, then we sleep and try again
+        await Future.delayed(_reconnectDelay);
+        _context = null; // reset _context, so next operation creates a new
+        return;
+      }
+      // If connecting was successful, then we await the connection being
+      // closed or error, and we reset _context.
+      // Any errors have already been returned if any commands were affect
+      // otherwise the error is just a broken TCP connection and we ignore it.
+      scheduleMicrotask(() async {
```

Comment:
Do you need this microtask here?

Reply:
Good, point I'm already in a micro task, I don't need to create a new one :)